lib/jxl/enc_ac_strategy.cc

```diff
@@ -580,196 +520,81 @@ void ProcessRectACS(PassesEncoderState* JXL_RESTRICT enc_state,
-        // In modes faster than Wombat mode, AC strategy replacement is not
-        // attempted: no need to estimate entropy.
-        if (cparams.speed_tier <= SpeedTier::kWombat) {
-          entropy =
-              EstimateEntropy(acs, (bx + ix) * 8, (by + iy) * 8, config,
-                              cmap_factors, block, scratch_space, quantized);
-          entropy *= entropy_adjust[i * 2 + 1];
-        }
-        // In modes faster than Hare mode, we don't use InitialQuantField -
-        // hence, we need to come up with quant field values.
-        if (cparams.speed_tier > SpeedTier::kHare &&
-            cparams.uniform_quant <= 0) {
-          // OPTIMIZE
-          float quant = 1.1f / (1.0f + max_delta_acs) / butteraugli_target;
-          for (size_t y = 0; y < cy; y++) {
-            for (size_t x = 0; x < cx; x++) {
-              config.SetQuant(bx + ix + x, by + iy + y, quant);
-            }
-          }
-        }
-        // Mark blocks as chosen and write to acs image.
-        ac_strategy->Set(bx + ix, by + iy, i);
-        for (size_t y = 0; y < cy; y++) {
-          for (size_t x = 0; x < cx; x++) {
-            chosen_mask[(y + iy) * 8 + x + ix] = 1;
-            entropy_estimate[(iy + y) * 8 + ix + x] = entropy;
-          }
-        }
-        break;
-      }
-    }
-  }
   // Do not try to replace ACS in modes faster than wombat mode.
   if (cparams.speed_tier > SpeedTier::kWombat) return;
-  // Iterate through the 32-block attempting to replace the current strategy.
-  // If replaced, repeat for the top-left new block and let the other ones be
-  // taken care of by future iterations.
-  uint8_t computed_mask[64] = {};
+  // First compute the best 8x8 transform for each square. Later, we do not
+  // experiment with different combinations, but only use the best of the 8x8s
+  // when DCT8X8 is specified in the tree search.
+  // 8x8 transforms have 10 variants, but every larger transform is just a DCT.
+  float entropy_estimate[64] = { 0 };
   for (size_t iy = 0; iy < rect.ysize(); iy++) {
     for (size_t ix = 0; ix < rect.xsize(); ix++) {
-      if (computed_mask[iy * 8 + ix]) continue;
-      uint8_t prev = AcStrategy::kNumValidStrategies;
-      while (prev != ac_strategy->ConstRow(by + iy)[bx + ix].RawStrategy()) {
-        prev = ac_strategy->ConstRow(by + iy)[bx + ix].RawStrategy();
-        MaybeReplaceACS(bx + ix, by + iy, config, cmap_factors, ac_strategy,
-                        entropy_adjust, entropy_estimate + (iy * 8 + ix), block,
-                        scratch_space, quantized);
-      }
-      AcStrategy acs = ac_strategy->ConstRow(by + iy)[bx + ix];
-      for (size_t y = 0; y < acs.covered_blocks_y(); y++) {
-        for (size_t x = 0; x < acs.covered_blocks_x(); x++) {
-          computed_mask[(iy + y) * 8 + ix + x] = 1;
-        }
+      float entropy = 0.0;
+      const uint8_t best_of_8x8s =
+	FindBest8x8Transform(8 * (bx + ix), 8 * (by + iy), config, cmap_factors, 
```

Comment:
What happens if we do this only if we don't merge the transform?

Reply:
I suppose it does help to know the estimated cost of the best 8x8 transform to compare it to the estimated cost of bigger ones, but yes, I would also expect that if you assume just the DCT8x8 initially (maybe underestimated a bit to take into account that there might be something better), then do the merging, and then find the best 8x8 for all remaining 8x8s, it shouldn't be much worse (and it should be faster, if merging happens).

Reply:
Small transforms work relatively well in high quality, but lower qualities benefit greatly from the interpolation characteristics of large transforms. I think that at high quality we lose 5 % and at low quality 25 % if we don't have large transforms.

Some additional work is needed for 8x8 dct to work optimally, like adjusting specially quantization so that the 2x2 corner is more likely to be preserved as it can being zeroed can lead to 8x8 blockiness to appear.